pinot-hadoop/src/main/java/org/apache/pinot/hadoop/job/SegmentPreprocessingJob.java

```diff
@@ -0,0 +1,479 @@
+    }
+
+    if (_enableResizing) {
+      fetchResizingConfig();
+      _logger.info("minimum number of output files: {}", _numberOfOutputFiles);
+    }
+
+    _logger.info("Initializing a pre-computing job");
+    Job job = Job.getInstance(_configuration);
+
+    job.getConfiguration().set(JobContext.JOB_NAME, this.getClass().getName());
+    // Turn this on to always firstly use class paths that user specifies.
+    job.getConfiguration().set(MRJobConfig.MAPREDUCE_JOB_USER_CLASSPATH_FIRST, "true");
+    // Turn this off since we don't need an empty file in the output directory
+    job.getConfiguration().set(FileOutputCommitter.SUCCESSFUL_JOB_OUTPUT_DIR_MARKER, "false");
+
+    job.setJarByClass(SegmentPreprocessingJob.class);
+
+    String hadoopTokenFileLocation = System.getenv("HADOOP_TOKEN_FILE_LOCATION");
+    if (hadoopTokenFileLocation != null) {
+      job.getConfiguration().set("mapreduce.job.credentials.binary", hadoopTokenFileLocation);
+    }
+
+    // Schema configs.
+    Schema schema = getSchema(inputDataPath.get(0));
+    _logger.info("Schema is: {}", schema.toString(true));
+
+    // Mapper configs.
+    job.setMapperClass(SegmentPreprocessingMapper.class);
+    job.setMapOutputKeyClass(AvroKey.class);
+    job.setMapOutputValueClass(AvroValue.class);
+    job.getConfiguration().setInt(JobContext.NUM_MAPS, inputDataPath.size());
+
+    // Reducer configs.
+    job.setReducerClass(SegmentPreprocessingReducer.class);
+    job.setOutputKeyClass(AvroKey.class);
+    job.setOutputValueClass(NullWritable.class);
+    if (_properties.getProperty(MAXIMUM_NUMBER_OF_RECORDS) != null) {
+      _logger.info("Setting {} to {}", MAXIMUM_NUMBER_OF_RECORDS, _properties.getProperty(MAXIMUM_NUMBER_OF_RECORDS));
+      job.getConfiguration().set(MAXIMUM_NUMBER_OF_RECORDS, _properties.getProperty(MAXIMUM_NUMBER_OF_RECORDS));
+    }
+    AvroMultipleOutputs.addNamedOutput(job, "avro", AvroKeyOutputFormat.class, schema);
+    AvroMultipleOutputs.setCountersEnabled(job, true);
+    LazyOutputFormat.setOutputFormatClass(job, AvroKeyOutputFormat.class);
+
+    // Input and output paths.
+    FileInputFormat.setInputPaths(job, _inputSegmentDir);
+    FileOutputFormat.setOutputPath(job, outputDir);
+
+    List<FileStatus> inputDataFiles = new ArrayList<>();
+    for (String inputPath : inputDataPath) {
+      Path inputPathPattern = new Path(inputPath);
+      FileStatus[] fileStatusArr = _fileSystem.globStatus(inputPathPattern);
+      for (FileStatus fileStatus : fileStatusArr) {
+        inputDataFiles.addAll(getDataFilesFromPath(_fileSystem, fileStatus.getPath()));
+      }
+    }
+    _logger.info("Total number of files to be pre-computed: {}", inputDataFiles.size());
+
+    // Set up mapper output key
+    Set<Schema.Field> fieldSet = new HashSet<>();
+
+    // Partition configs.
+    int numReduceTasks = (_numberOfPartitions != 0) ? _numberOfPartitions : inputDataFiles.size();
+    if (_partitionColumn != null) {
```

Comment:
do we have diff partition fns that we support? Or only the murmur one? IF only murmur, why do we allow this to be configurable?

also, might be nice to do extra error checking here, like make sure partition column and num partitions set and not partition if only one is set, for examplee

Reply:
It's not just murmur. Pinot also supports other functions like modulo, default(hashcode). I'll add the error check.